base_layer/core/src/transactions/transaction_components/encrypted_openings_x.rs

```diff
@@ -0,0 +1,227 @@
+    fn from(err: Error) -> Self {
+        Self::EncryptionFailed(err)
+    }
+}
+
+impl EncryptedOpeningsX {
+    const TAG: &'static [u8] = b"TARI_AAD_VALUE_AND_MASK";
+
+    /// Encrypt the value and mask (with fixed length) using XChaCha20-Poly1305 with a secure random nonce
+    pub fn encrypt_openings(
+        encryption_key: &PrivateKey,
+        commitment: &Commitment,
+        value: MicroTari,
+        mask: &PrivateKey,
+    ) -> Result<EncryptedOpeningsX, EncryptedOpeningsErrorX> {
+        let mut openings = value.as_u64().to_le_bytes().to_vec();
+        openings.append(&mut mask.to_vec());
+        let aead_payload = Payload {
+            msg: openings.as_slice(),
+            aad: Self::TAG,
+        };
+
+        // Produce a secure random nonce
+        let mut nonce = [0u8; size_of::<XNonce>()];
+        OsRng.fill_bytes(&mut nonce);
+        let nonce_ga = XNonce::from_slice(&nonce);
+
+        let aead_key = kdf_aead(encryption_key, commitment);
+        let cipher = XChaCha20Poly1305::new(GenericArray::from_slice(aead_key.reveal()));
+        let mut ciphertext = cipher.encrypt(nonce_ga, aead_payload)?;
+        let mut ciphertext_integral_nonce = nonce.to_vec();
+        ciphertext_integral_nonce.append(&mut ciphertext);
+
+        EncryptedOpeningsX::from_bytes(ciphertext_integral_nonce.as_slice())
+    }
+
+    /// Authenticate and decrypt the value and mask
+    pub fn decrypt_openings(
+        encryption_key: &PrivateKey,
+        commitment: &Commitment,
+        encrypted_openings: &EncryptedOpeningsX,
+    ) -> Result<(MicroTari, PrivateKey), EncryptedOpeningsErrorX> {
+        // Extract the nonce and ciphertext
+        let binding = encrypted_openings.as_byte_vector();
+        let (nonce, ciphertext) = binding.split_at(size_of::<XNonce>());
+        let nonce_ga = XNonce::from_slice(nonce);
+
+        let aead_key = kdf_aead(encryption_key, commitment);
+        let cipher = XChaCha20Poly1305::new(GenericArray::from_slice(aead_key.reveal()));
+        let aead_payload = Payload {
+            msg: ciphertext,
+            aad: Self::TAG,
+        };
+        let decrypted_bytes = cipher.decrypt(nonce_ga, aead_payload)?;
+        let mut value_bytes = [0u8; VALUE_SIZE];
+        value_bytes.clone_from_slice(&decrypted_bytes[0..VALUE_SIZE]);
+        let mut mask_bytes = [0u8; KEY_SIZE];
+        mask_bytes.clone_from_slice(&decrypted_bytes[VALUE_SIZE..VALUE_SIZE + KEY_SIZE]);
+        Ok((
+            u64::from_le_bytes(value_bytes).into(),
+            PrivateKey::from_bytes(&mask_bytes)?,
+        ))
+    }
+}
+
+// Generate a ChaCha20-Poly1305 key from a private key and commitment using Blake2b
+fn kdf_aead(encryption_key: &PrivateKey, commitment: &Commitment) -> EncryptedOpeningsKey {
+    let mut aead_key = EncryptedOpeningsKey::from(SafeArray::default());
+    DomainSeparatedHasher::<Blake256, TransactionKdfDomain>::new_with_label("encrypted_value_and_mask")
```

Comment:
Recommend a distinct domain separator for this, since the fixed-nonce version is technically a different algorithm.